### PR TITLE
[DO NOT MERGE] [closure-specializer] Support specialization of closures with address parameters

### PR DESCRIPTION
**Pull request by swiftix**
This is a more advanced version of #9987. But it is not quite correct when it comes to insertion of argument copies around the closure calls yet.

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please smoke benchmark

File: lib/SILOptimizer/IPO/ClosureSpecializer.cpp [hunk 1]
```diff
@@ -338,17 +417,40 @@ static void rewriteApplyInst(const CallSiteDescriptor &CSDesc,
...
-
-      // Emit the retain that matches the captured argument by the partial_apply
-      // in the callee that is consumed by the partial_apply.
-      Builder.setInsertionPoint(AI.getInstruction());
-      Builder.createRetainValue(Closure->getLoc(), Arg, Builder.getDefaultAtomicity());
+      //if (ArgTy.isObject()) //???
+      CSDesc.extendArgumentLifetime(Arg, ArgConvention);
     } else {
-      Builder.createRetainValue(Closure->getLoc(), Arg, Builder.getDefaultAtomicity());
+      if (ArgTy.isObject())
```
Comment (shahzadlone) [inline lib/SILOptimizer/IPO/ClosureSpecializer.cpp]: perhaps ```braces``` for the ```if``` to avoid mistakes.

**Comment by CodaFi** [comment]
It seems like SIL Opaque values are the answer here.  Closing this out.

**Pull request closed without merging**